c-deps/libroach/stack_trace.cc

```diff
@@ -0,0 +1,271 @@
+  return pids;
+}
+
+uint64_t BlockedSignals(pid_t tid) {
+  const std::string path = "/proc/" + std::to_string(tid) + "/status";
+  int fd;
+  do {
+    fd = open(path.c_str(), O_RDONLY);
+  } while (fd < 0 && errno == EINTR);
+  if (fd < -1) {
+    return 0;
+  }
+  std::string data;
+  for (;;) {
+    char buf[1024];
+    int n;
+    do {
+      n = read(fd, buf, sizeof(buf));
+    } while (n < 0 && errno == EINTR);
+    if (n < 0) {
+      close(fd);
+      return 0;
+    }
+    if (n == 0) {
+      break;
+    }
+    data.append(buf, n);
+  }
+  close(fd);
+
+  const std::string needle("SigBlk:");
+  size_t pos = data.find(needle);
+  if (pos == data.npos) {
+    return 0;
+  }
+  data = data.substr(pos + needle.size());
+  return stoull(data, nullptr, 16);
+}
+
+void InternalHandler(int signum, siginfo_t* siginfo, void* ucontext) {
+  // Ignore signals that were sent by an external process. The
+  // stacktrace signal handler is intended only for signals we send to
+  // ourselves.
+  if (siginfo->si_pid != getpid()) {
+    return;
+  }
+  auto stack = reinterpret_cast<ThreadStack*>(siginfo->si_value.sival_ptr);
+  if (stack == nullptr) {
+    return;
+  }
+  stack->depth = backtrace(stack->addr, kMaxDepth);
+  stack->Ack();
+}
+
+int SignalThread(pid_t pid, pid_t tid, uid_t uid, int signum, sigval payload) {
+  // Similar to pthread_sigqueue(), but usable with a tid since we
+  // don't have a pthread_t.
+  siginfo_t info;
+  memset(&info, 0, sizeof(info));
+  info.si_signo = signum;
+  info.si_code = SI_QUEUE;
+  info.si_pid = pid;
+  info.si_uid = uid;
+  info.si_value = payload;
+  return syscall(SYS_rt_tgsigqueueinfo, pid, tid, signum, &info);
+}
+
+bool InitThreadDumper() {
+  struct sigaction action;
+  memset(&action, 0, sizeof(action));
+  action.sa_sigaction = InternalHandler;
+  // Set SA_RESTART so that supported syscalls are automatically restarted if
+  // interrupted by the stacktrace collection signal.
+  action.sa_flags = SA_ONSTACK | SA_RESTART | SA_SIGINFO;
+  return sigaction(kStackTraceSignal, &action, nullptr) == 0;
+}
+
+int64_t NowMillis() {
+  timeval tv;
+  gettimeofday(&tv, NULL);
+  return (tv.tv_sec * 1000) + (tv.tv_usec / 1000);
+}
+
+}  // namespace
+
+std::string DumpThreadStacks() {
+  auto tids = ListThreads();
+  InitThreadDumper();
+
+  // Create a pipe on which threads can send acks after they finish
+  // writing their stacktrace.
+  int pipe_fd[2];
+  if (pipe(pipe_fd) == -1) {
+    return "unable to create pipe";
+  }
+
+  // Signal all threads to write their stack trace in a pre-allocated
+  // area. Note that some threads might have died by now, so
+  // signalling them will fail.
+  std::vector<std::unique_ptr<ThreadStack>> stacks;
+  const auto pid = getpid();
+  const auto uid = getuid();
+  for (auto tid : tids) {
+    const uint64_t blocked = BlockedSignals(tid);
+    if ((blocked & (1ULL << kStackTraceSignal)) != 0) {
+      // The thread is blocking receipt of our signal, so don't bother
```

Comment:
What does this mean (i.e. when would a thread not accept our signal) and does that mean we're just omitting it? Is that ok?